core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1483,21 +1512,53 @@ class KafkaApis(val requestChannel: RequestChannel,
     val transactionalId = addPartitionsToTxnRequest.transactionalId
     val partitionsToAdd = addPartitionsToTxnRequest.partitions
 
-    // Send response callback
-    def sendResponseCallback(error: Errors): Unit = {
-      def createResponse(throttleTimeMs: Int): AbstractResponse = {
-        val responseBody: AddPartitionsToTxnResponse = new AddPartitionsToTxnResponse(throttleTimeMs, error)
-        trace(s"Completed $transactionalId's AddPartitionsToTxnRequest with partitions $partitionsToAdd: errors: $error from client ${request.header.clientId}")
-        responseBody
+    if(!authorize(request.session, Write, new Resource(ProducerTransactionalId, transactionalId)))
+
+      sendResponseMaybeThrottle(request, (throttleTimeMs: Int) => new AddPartitionsToTxnResponse(throttleTimeMs,
+        partitionsToAdd.asScala.map{tp => (tp, Errors.TRANSACTIONAL_ID_AUTHORIZATION_FAILED)}.toMap.asJava))
+    else {
+      val internalTopics = partitionsToAdd.asScala.filter {tp => org.apache.kafka.common.internals.Topic.isInternal(tp.topic())}
+
+      val (existingAndAuthorizedForDescribeTopics, nonExistingOrUnauthorizedForDescribeTopics) =
+        partitionsToAdd.asScala.partition { tp =>
+          authorize(request.session, Describe, new Resource(Topic, tp.topic)) && metadataCache.contains(tp.topic)
+        }
+
+      val (authorizedRequestInfo, unauthorizedForWriteRequestInfo) = existingAndAuthorizedForDescribeTopics.partition {
+        tp => authorize(request.session, Write, new Resource(Topic, tp.topic))
+      }
+
+      if (nonExistingOrUnauthorizedForDescribeTopics.nonEmpty
+        || unauthorizedForWriteRequestInfo.nonEmpty
+        || internalTopics.nonEmpty) {
+
+        val partitionErrors = unauthorizedForWriteRequestInfo.map { tp => (tp, Errors.TOPIC_AUTHORIZATION_FAILED) }.toMap ++
```

Comment:
So it seems like we only send back the partitions which failed. I think that's reasonable since returning Errors.NONE for the partitions which didn't fail would be misleading.  Can we add a comment which explains that this behavior is intentional?